File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/patcher/PatcherUtils.java
/*
 * Copyright (C) 2014  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.patcher;

import android.content.Context;
import android.os.Environment;
import android.util.Log;

import com.github.chenxiaolong.dualbootpatcher.BuildConfig;
import com.github.chenxiaolong.dualbootpatcher.FileUtils;
import com.github.chenxiaolong.dualbootpatcher.R;
import com.github.chenxiaolong.dualbootpatcher.RomUtils;
import com.github.chenxiaolong.dualbootpatcher.ThreadUtils;
import com.github.chenxiaolong.dualbootpatcher.nativelib.LibMbDevice.Device;
import com.github.chenxiaolong.dualbootpatcher.nativelib.LibMbp.PatcherConfig;
import com.github.chenxiaolong.dualbootpatcher.nativelib.libmiscstuff.LibMiscStuff;

import org.apache.commons.io.Charsets;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

public class PatcherUtils {
    private static final String TAG = PatcherUtils.class.getSimpleName();
    private static final String FILENAME = "data-%s.tar.xz";
    private static final String DIRNAME = "data-%s";

    private static final String PREFIX_DATA_SLOT = "data-slot-";
    private static final String PREFIX_EXTSD_SLOT = "extsd-slot-";

    public static final String PATCHER_ID_MULTIBOOTPATCHER = "MultiBootPatcher";
    public static final String PATCHER_ID_ODINPATCHER = "OdinPatcher";

    private static boolean sInitialized;

    private static Device[] sDevices;
    private static Device sCurrentDevice;

    private static String sTargetFile;
    private static String sTargetDir;

    private static InstallLocation[] sInstallLocations;

    static {
        String version = BuildConfig.VERSION_NAME.split("-")[0];
        sTargetFile = String.format(FILENAME, version);
        sTargetDir = String.format(DIRNAME, version);
    }

    private static File getTargetFile(Context context) {
        return new File(context.getCacheDir() + File.separator + sTargetFile);
    }

    public static File getTargetDirectory(Context context) {
        return new File(context.getFilesDir() + File.separator + sTargetDir);
    }

    public synchronized static void initializePatcher(Context context) {
        if (!sInitialized) {
            extractPatcher(context);
            sInitialized = true;
        }
    }

    public static PatcherConfig newPatcherConfig(Context context) {
        PatcherConfig pc = new PatcherConfig();
        pc.setDataDirectory(getTargetDirectory(context).getAbsolutePath());
        pc.setTempDirectory(context.getCacheDir().getAbsolutePath());
        return pc;
    }

    public synchronized static Device[] getDevices(Context context) {
        ThreadUtils.enforceExecutionOnNonMainThread();
        initializePatcher(context);

        if (sDevices == null) {
            File path = new File(getTargetDirectory(context) + File.separator + "devices.json");
            try {
                String json = org.apache.commons.io.FileUtils.readFileToString(
                        path, Charsets.UTF_8);

                Device[] devices = Device.newListFromJson(json);
                ArrayList<Device> validDevices = new ArrayList<>();

                if (devices != null) {
                    for (Device d : devices) {
                        if (d.validate() == 0) {
                            validDevices.add(d);
                        }
                    }
                }

                if (!validDevices.isEmpty()) {
                    sDevices = validDevices.toArray(new Device[validDevices.size()]);
                }
            } catch (IOException e) {
                Log.w(TAG, "Failed to read " + path, e);
            }
        }

        return sDevices;
    }

    public synchronized static Device getCurrentDevice(Context context) {
        ThreadUtils.enforceExecutionOnNonMainThread();

        if (sCurrentDevice == null) {
            String realCodename = RomUtils.getDeviceCodename(context);
            Device[] devices = getDevices(context);

            if (devices != null) {
                outer:
                for (Device d : devices) {
                    for (String codename : d.getCodenames()) {
                        if (realCodename.equals(codename)) {
                            sCurrentDevice = d;
                            break outer;
                        }
                    }
                }
            }
        }

        return sCurrentDevice;
    }

    public synchronized static void extractPatcher(Context context) {
        for (File d : context.getCacheDir().listFiles()) {
            if (d.getName().startsWith("DualBootPatcherAndroid")
                    || d.getName().startsWith("tmp")
                    || d.getName().startsWith("data-")) {
                org.apache.commons.io.FileUtils.deleteQuietly(d);
            }
        }
        for (File d : context.getFilesDir().listFiles()) {
            if (d.isDirectory()) {
                for (File t : d.listFiles()) {
                    if (t.getName().contains("tmp")) {
                        org.apache.commons.io.FileUtils.deleteQuietly(t);
                    }
                }
            }
        }

        File targetFile = getTargetFile(context);
        File targetDir = getTargetDirectory(context);

        if (BuildConfig.BUILD_TYPE.equals("debug") || !targetDir.exists()) {
            try {
                FileUtils.extractAsset(context, sTargetFile, targetFile);
            } catch (IOException e) {
                throw new IllegalStateException("Failed to extract data archive from assets", e);
            }

            // Remove all previous files
            for (File d : context.getFilesDir().listFiles()) {
                org.apache.commons.io.FileUtils.deleteQuietly(d);
            }

            try {
                LibMiscStuff.extractArchive(targetFile.getAbsolutePath(),
                        context.getFilesDir().getAbsolutePath());
            } catch (IOException e) {
                throw new IllegalStateException("Failed to extract data archive", e);
            }

            // Delete archive
            targetFile.delete();
        }
    }

    public static class InstallLocation {
        public String id;
        public String name;
        public String description;
    }

    public static InstallLocation[] getInstallLocations(Context context) {
        if (sInstallLocations == null) {
            ArrayList<InstallLocation> locations = new ArrayList<>();

            InstallLocation location = new InstallLocation();
            location.id = "primary";
            location.name = context.getString(R.string.install_location_primary_upgrade);
            location.description = context.getString(R.string.install_location_primary_upgrade_desc);

            locations.add(location);

            location = new InstallLocation();
            location.id = "dual";
            location.name = context.getString(R.string.secondary);
            location.description = String.format(context.getString(R.string.install_location_desc),
                    "/system/multiboot/dual");
            locations.add(location);

            for (int i = 1; i <= 3; i++) {
                location = new InstallLocation();
                location.id = "multi-slot-" + i;
                location.name = String.format(context.getString(R.string.multislot), i);
                location.description = String.format(context.getString(R.string.install_location_desc),
                        "/cache/multiboot/multi-slot-" + i);
                locations.add(location);
            }

            sInstallLocations = locations.toArray(new InstallLocation[locations.size()]);
        }

        return sInstallLocations;
    }

    public static InstallLocation[] getNamedInstallLocations(Context context) {
        ThreadUtils.enforceExecutionOnNonMainThread();

        Log.d(TAG, "Looking for named ROMs");

        File dir = new File(Environment.getExternalStorageDirectory()
                + File.separator + "MultiBoot");

        ArrayList<InstallLocation> locations = new ArrayList<>();

        File[] files = dir.listFiles();
        if (files != null) {
            for (File f : files) {
                String name = f.getName();

                if (name.startsWith("data-slot-") && !name.equals("data-slot-")) {
                    Log.d(TAG, "- Found data-slot: " + name.substring(10));
                    locations.add(getDataSlotInstallLocation(context, name.substring(10)));
                } else if (name.startsWith("extsd-slot-") && !name.equals("extsd-slot-")) {
                    Log.d(TAG, "- Found extsd-slot: " + name.substring(11));
                    locations.add(getExtsdSlotInstallLocation(context, name.substring(11)));
                }
            }
        } else {
            Log.e(TAG, "Failed to list files in: " + dir);
        }

        return locations.toArray(new InstallLocation[locations.size()]);
    }

    public static InstallLocation getDataSlotInstallLocation(Context context, String dataSlotId) {
        InstallLocation location = new InstallLocation();
        location.id = getDataSlotRomId(dataSlotId);
        location.name = String.format(context.getString(R.string.dataslot), dataSlotId);
        location.description = context.getString(R.string.install_location_desc,
                "/data/multiboot/data-slot-" + dataSlotId);
        return location;
    }

    public static InstallLocation getExtsdSlotInstallLocation(Context context, String extsdSlotId) {
        InstallLocation location = new InstallLocation();
        location.id = getExtsdSlotRomId(extsdSlotId);
        location.name = String.format(context.getString(R.string.extsdslot), extsdSlotId);
        location.description = context.getString(R.string.install_location_desc,
                "[External SD]/multiboot/extsd-slot-" + extsdSlotId);
        return location;
    }

    public static String getDataSlotRomId(String dataSlotId) {
        return PREFIX_DATA_SLOT + dataSlotId;
    }

    public static String getExtsdSlotRomId(String extsdSlotId) {
        return PREFIX_EXTSD_SLOT + extsdSlotId;
    }

    public static boolean isDataSlotRomId(String romId) {
        return romId.startsWith(PREFIX_DATA_SLOT);
    }

    public static boolean isExtsdSlotRomId(String romId) {
        return romId.startsWith(PREFIX_EXTSD_SLOT);
    }

    public static String getDataSlotIdFromRomId(String romId) {
        if (isDataSlotRomId(romId)) {
            return romId.substring(PREFIX_DATA_SLOT.length());
        } else {
            return null;
        }
    }

    public static String getExtsdSlotIdFromRomId(String romId) {
        if (isExtsdSlotRomId(romId)) {
            return romId.substring(PREFIX_EXTSD_SLOT.length());
        } else {
            return null;
        }
    }
}

File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/patcher/PatcherUtils.kt
/*
 * Copyright (C) 2014  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.patcher

import android.content.Context
import android.os.Environment
import android.util.Log
import com.github.chenxiaolong.dualbootpatcher.BuildConfig
import com.github.chenxiaolong.dualbootpatcher.FileUtils
import com.github.chenxiaolong.dualbootpatcher.R
import com.github.chenxiaolong.dualbootpatcher.RomUtils
import com.github.chenxiaolong.dualbootpatcher.ThreadUtils
import com.github.chenxiaolong.dualbootpatcher.nativelib.LibMbDevice.Device
import com.github.chenxiaolong.dualbootpatcher.nativelib.LibMbPatcher.PatcherConfig
import com.github.chenxiaolong.dualbootpatcher.nativelib.libmiscstuff.LibMiscStuff
import java.io.File
import java.io.IOException
import java.util.ArrayList

object PatcherUtils {
    private val TAG = PatcherUtils::class.java.simpleName
    private const val FILENAME = "data-%s.tar.xz"
    private const val DIRNAME = "data-%s"

    const val PATCHER_ID_ZIPPATCHER = "ZipPatcher"
    const val PATCHER_ID_ODINPATCHER = "OdinPatcher"

    private val PRIMARY_SLOT = InstallLocation("primary",
            R.string.primary, emptyArray(),
            R.string.install_location_desc, arrayOf("/system"))

    private val SECONDARY_SLOT = InstallLocation("dual",
            R.string.secondary, emptyArray(),
            R.string.install_location_desc, arrayOf("/system/multiboot/dual"))

    private val MULTI_SLOT_TEMPLATE = TemplateLocation("multi-slot-",
            0, emptyArray(),
            R.string.multislot, arrayOf(TemplateLocation.PLACEHOLDER_SUFFIX),
            R.string.install_location_desc,
                    arrayOf("/cache/multiboot/${TemplateLocation.PLACEHOLDER_ID}"))

    private val DATA_SLOT_TEMPLATE = TemplateLocation("data-slot-",
            R.string.install_location_data_slot, emptyArray(),
            R.string.dataslot, arrayOf(TemplateLocation.PLACEHOLDER_SUFFIX),
            R.string.install_location_desc,
                    arrayOf("/data/multiboot/${TemplateLocation.PLACEHOLDER_ID}"))

    private val EXTSD_SLOT_TEMPLATE = TemplateLocation("extsd-slot-",
            R.string.install_location_extsd_slot, emptyArray(),
            R.string.extsdslot, arrayOf(TemplateLocation.PLACEHOLDER_SUFFIX),
            R.string.install_location_desc,
                    arrayOf("[External SD]/multiboot/${TemplateLocation.PLACEHOLDER_ID}"))

    private var initialized: Boolean = false

    private var devices: List<Device>? = null
    private var currentDevice: Device? = null

    private var targetFile: String
    private var targetDir: String

    var installLocations: Array<InstallLocation>
    var templateLocations: Array<TemplateLocation>

    init {
        val version = BuildConfig.VERSION_NAME.split("-")[0]
        targetFile = String.format(FILENAME, version)
        targetDir = String.format(DIRNAME, version)

        installLocations = arrayOf(
                PRIMARY_SLOT,
                SECONDARY_SLOT,
                MULTI_SLOT_TEMPLATE.toInstallLocation("1"),
                MULTI_SLOT_TEMPLATE.toInstallLocation("2"),
                MULTI_SLOT_TEMPLATE.toInstallLocation("3"))

        templateLocations = arrayOf(DATA_SLOT_TEMPLATE, EXTSD_SLOT_TEMPLATE)
    }

    private fun getTargetFile(context: Context): File {
        return File(context.cacheDir, targetFile)
    }

    fun getTargetDirectory(context: Context): File {
        return File(context.filesDir, targetDir)
    }

    @Synchronized
    fun initializePatcher(context: Context) {
        if (!initialized) {
            extractPatcher(context)
            initialized = true
        }
    }

    fun newPatcherConfig(context: Context): PatcherConfig {
        val pc = PatcherConfig()
        pc.dataDirectory = getTargetDirectory(context).absolutePath
        pc.tempDirectory = context.cacheDir.absolutePath
        return pc
    }

    @Synchronized
    fun getDevices(context: Context): List<Device>? {
        ThreadUtils.enforceExecutionOnNonMainThread()
        initializePatcher(context)

        if (devices == null) {
            val path = File(getTargetDirectory(context), "devices.json")
            try {
                val json = path.readText()

                val validDevices = ArrayList<Device>()
                Device.newListFromJson(json)?.filterTo(validDevices) { it.validate() == 0L }

                if (!validDevices.isEmpty()) {
                    this.devices = validDevices
                }
            } catch (e: IOException) {
                Log.w(TAG, "Failed to read $path", e)
            }
        }

        return devices
    }

    @Synchronized
    fun getCurrentDevice(context: Context): Device? {
        ThreadUtils.enforceExecutionOnNonMainThread()

        if (currentDevice == null) {
            val devices = getDevices(context)
            if (devices != null) {
                currentDevice = getCurrentDevice(context, devices)
            }
        }

        return currentDevice
    }

    fun getCurrentDevice(context: Context, devices: Collection<Device>): Device? {
        val realCodename = RomUtils.getDeviceCodename(context)

        for (d in devices) {
            for (codename in d.codenames!!) {
                if (realCodename == codename) {
                    return d
                }
            }
        }

        return null
    }

    @Synchronized
    fun extractPatcher(context: Context) {
        context.cacheDir.listFiles()
                .filter {
                    it.name.startsWith("DualBootPatcherAndroid")
                            || it.name.startsWith("tmp")
                            || it.name.startsWith("data-")
                }
                .forEach { it.deleteRecursively() }
        context.filesDir.listFiles()
                .filter { it.isDirectory }
                .forEach {
                    it.listFiles()
                            .filter { it.name.contains("tmp") }
                            .forEach { it.deleteRecursively() }
                }

        val targetFile = getTargetFile(context)
        val targetDir = getTargetDirectory(context)

        if (BuildConfig.BUILD_TYPE == "debug" || !targetDir.exists()) {
            try {
                FileUtils.extractAsset(context, this.targetFile, targetFile)
            } catch (e: IOException) {
                throw IllegalStateException("Failed to extract data archive from assets", e)
            }

            // Remove all previous files
            context.filesDir.listFiles().forEach {
                it.deleteRecursively()
            }

            try {
                LibMiscStuff.extractArchive(targetFile.absolutePath, context.filesDir.absolutePath)
            } catch (e: IOException) {
                throw IllegalStateException("Failed to extract data archive", e)
            }

            // Delete archive
            targetFile.delete()
        }
    }

    fun getInstalledTemplateLocations(): Array<InstallLocation> {
        ThreadUtils.enforceExecutionOnNonMainThread()

        Log.d(TAG, "Looking for existing template install locations")

        val dir = File(Environment.getExternalStorageDirectory(), "MultiBoot")

        val locations = ArrayList<InstallLocation>()

        val files = dir.listFiles()
        if (files != null) {
            files.map { it.name }.forEach { filename ->
                for (template in templateLocations) {
                    if (filename.startsWith(template.prefix) && filename != template.prefix) {
                        val loc = template.toInstallLocation(filename.substring(template.prefix.length))
                        Log.d(TAG, "- Found ${loc.id}")
                        locations.add(loc)
                        break
                    }
                }
            }
        } else {
            Log.e(TAG, "Failed to list files in: $dir")
        }

        return locations.toTypedArray()
    }

    fun getInstallLocationFromRomId(romId: String): InstallLocation? {
        return installLocations.firstOrNull { it.id == romId }
                ?: templateLocations
                        .firstOrNull { romId.startsWith(it.prefix) }
                        ?.let { it.toInstallLocation(romId.substring(it.prefix.length)) }
    }
}
